### DISCUSSION TITLE
Ways to throw custom error messages

### DISCUSSION THREAD

**Discussion by tommypepsi** [Shopify/function-examples#247] [Feedback: Cart and Checkout Validation API]
When adding the cart validation rule, the user has the choice to block the checkout or let it continue in case of errors. The error is really generic and just tells the customer to contact the store owner. I'm pretty sure this will cause a lot of confusion as there is no indication as to what might be causing this error. 

One thing I would like to be able to do is throw an error with a customized error message (that would show in the checkout) that would also respect the choice of the merchant to block the checkout or not. 

An example is our app depends on data from the frontend to be passed to the order attributes, if this information is missing (due to the app failing) I would like to block the checkout if this was the choice of the merchant when he added the rule or continue if not.

I thought of two possible ways for that:
- By literally throwing an error with a message that could be shown on the front end:
```javascript
if (!input.cart.validationKey) {
    throw new Error("Limits verification failed.");
}
```
I'm guessing we would need to define a specific way to throw the error to not show unwanted error message to the front end.

- By having access to the choice of the merchant in the input of the function:

```graphql
query Input {
  cart {
    ...
  }
  blockCheckoutOnError
}
```
```javascript
if (!input.cart.validationKey) {
  if(input.blockCheckoutOnError) {
    return {
      errors: [
        {
          localizedMessage: "Limits verification failed.",
          target: "cart",
        }
      ]
    }
  }
  else {
    return { errors: [] };
  }
}
```

I'm guessing we could add an option within our app with a metafield for this but since the merchant is already making this choice when adding the rule I think it would be redundant to ask this again.

**Comment by aledalgrande** [comment]
Hey @tommypepsi

> The error is really generic and just tells the customer to contact the store owner. 

That error is only shown when a function crashes while running. If you see that message while running the function in your dev store, you should be able to go check your extension in Partners and see what kind of exception made the function fail.

We already support outputting [errors](https://shopify.dev/docs/api/functions/reference/cart-checkout-validation/graphql/functionresult) from functions, like in [this example](https://shopify.dev/docs/apps/checkout/validation/server-side#step-1-create-the-cart-and-checkout-validation-function).

Happy to hear suggestions for improvements, if the documentation was hard to find/confusing.

**Reply by tommypepsi** [reply]
Hi @aledalgrande, yes I understand that but I mean I would like to be able to ouput optional errors based on the choice that the merchant chose here:
![image](https://github.com/Shopify/function-examples/assets/926715/f5d65748-e109-471a-8711-2e0a48b830c8)

For example if our function is missing information to complete the validation I would like to show an error if the merchant decided to block the checkout in case of errors or continue if he selected the continue checkout. I could throw an error but the default message is too generic for that.

Like I said, one way to do this could be to be able to retrieve that choice from the function input. Or another way I just thought could be a new option in the return like that:
```javascript
{
  errors: [
    {
      localizedMessage: "Limits verification failed.",
      target: "cart",
      optional: true
    }
  ]
}
```

where optional would respect the choice to block or continue the checkout.

**Reply by aledalgrande** [reply]
Thanks for going over your explanation again @tommypepsi. We added to our roadmap a new attribute for the input graph that will help you understand which setting the merchant chose, so you can decide if to display an error.
